x/pairing/keeper/scores/score.go

```diff
@@ -0,0 +1,230 @@
+import (
+	"fmt"
+	"math/big"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	"github.com/lavanet/lava/utils"
+	epochstoragetypes "github.com/lavanet/lava/x/epochstorage/types"
+	planstypes "github.com/lavanet/lava/x/plans/types"
+	tendermintcrypto "github.com/tendermint/tendermint/crypto"
+)
+
+var (
+	uniformStrategy ScoreStrategy
+	allReqNames     []string
+)
+
+// TODO: currently we'll use weight=1 for all reqs. In the future, we'll get it from policy
+func init() {
+	// gather all req names to a list
+	allReqNames = append(allReqNames, stake_req_name)
+
+	// init strategy
+	uniformStrategy = make(ScoreStrategy)
+	for _, reqName := range allReqNames {
+		uniformStrategy[reqName] = 1
+	}
+}
+
+// get the overall requirements from the policy and assign slots that'll fulfil them
+// TODO: this function should be changed in the future since it only supports stake reqs
+func CalcSlots(policy planstypes.Policy, minStake sdk.Int) []*PairingSlot {
+	// init slot array (should be as the number of providers to pair)
+	slots := make([]*PairingSlot, policy.MaxProvidersToPair)
+
+	// all slots should consider the stake, so we init them with stakeReq
+	stakeReq := StakeReq{MinStake: minStake}
+	slotReqs := map[string]ScoreReq{stakeReq.GetName(): stakeReq}
+	for i := range slots {
+		slots[i] = NewPairingSlot(slotReqs)
+	}
+
+	return slots
+}
+
+// group the slots
+func GroupSlots(slots []*PairingSlot) []PairingSlotGroup {
+	slotGroups := []PairingSlotGroup{}
+	if len(slots) == 0 {
+		utils.LavaFormatError("no slots", sdkerrors.ErrLogic)
+		return slotGroups
+	}
+
+	for k := range slots {
+		foundGroup := false
+		for i := range slotGroups {
+			diff := slots[k].Diff(slotGroups[i].Slot)
+			if len(diff.Reqs) == 0 {
+				slotGroups[i].Count += 1
+				foundGroup = true
+				break
+			}
+		}
+
+		if !foundGroup {
+			newGroup := NewPairingSlotGroup(slots[k])
+			slotGroups = append(slotGroups, *newGroup)
+		}
+	}
+
+	return slotGroups
+}
+
+// TODO: currently we'll use weight=1 for all reqs. In the future, we'll get it from policy
+func GetStrategy() ScoreStrategy {
+	return uniformStrategy
+}
+
+// calculates the final pairing score for all slot groups (with strategy)
+// we calculate only the diff between the current and previous slot groups
+func CalcPairingScore(scores []*PairingScore, strategy ScoreStrategy, diffSlot *PairingSlot, minStake sdk.Int) error {
+	// calculate the score for each req for each provider
+	for _, req := range diffSlot.Reqs {
+		reqName := req.GetName()
+		weight, ok := strategy[reqName]
+		if !ok {
+			return utils.LavaFormatError("req not in strategy", sdkerrors.ErrKeyNotFound,
+				utils.Attribute{Key: "req_name", Value: reqName})
+		}
+
+		for _, score := range scores {
+			// normalize stake so we won't overflow the score result (uint64)
+			newScoreComp := req.Score(*score.Provider, weight)
+
+			// divide by previous score component (if exists) and multiply by new score
```

Comment:
why is this?

Reply:
basically to avoid unnecessary calculation. I recalculate the score components of diffSlot so this is the way to update the provider's existing score according to specific components only